elan/rpc/retry.go

```diff
@@ -0,0 +1,98 @@
+package rpc
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"gocloud.dev/blob"
+	"gocloud.dev/gcerrors"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// retryBucket implements a retry logic on top of all bucket methods that can
+// return an error. Those methods are retried only on internal errors that are
+// not caused by a context cancellation.
+type retryBucket struct {
+	bucket
+	retries int
+}
+
+func (b retryBucket) NewRangeReader(ctx context.Context, key string, offset, length int64, opts *blob.ReaderOptions) (io.ReadCloser, error) {
+	var (
```

Comment:
I feel like generics might provide a nicer way of sharing the implementation of these?